### PR TITLE
expression: support `str_to_date` push to TiFlash (#25095)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #25095 to release-5.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/25148
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/25148:release-5.0-090892822b17
```

---

Signed-off-by: JaySon-Huang <jayson.hjs@gmail.com>

### What problem does this PR solve?

Issue Number: relate to https://github.com/pingcap/tics/issues/1942

Problem Summary:
Function: push `str_to_date` to TiFlash.

### What is changed and how it works?

Add push down support for `str_to_date`

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Support pushing down the `str_to_date` functions to TiFlash

### CODE REVIEW

File: expression/expr_to_pb_test.go [hunk 1]
```diff
@@ -724,6 +724,60 @@ func (s *testEvaluatorSuite) TestExprPushDownToFlash(c *C) {
...
+	c.Assert(function.(*ScalarFunction).Function.PbCode(), Equals, tipb.ScalarFuncSig_UnixTimestampDec)
+	exprs = append(exprs, function)
+
+	// StrToDateDateTime
+	function, err = NewFunction(mock.NewContext(), ast.StrToDate, types.NewFieldType(mysql.TypeDatetime), stringColumn, stringColumn)
+	c.Assert(err, IsNil)
+	c.Assert(function.(*ScalarFunction).Function.PbCode(), Equals, tipb.ScalarFuncSig_StrToDateDatetime)
+	exprs = append(exprs, function)
+
+>>>>>>> 090892822... expression: support `str_to_date` push to TiFlash (#25095)
```
Comment (JaySon-Huang) [inline expression/expr_to_pb_test.go:780]:
```suggestion
	// StrToDateDateTime
	function, err = NewFunction(mock.NewContext(), ast.StrToDate, types.NewFieldType(mysql.TypeDatetime), stringColumn, stringColumn)
	c.Assert(err, IsNil)
	c.Assert(function.(*ScalarFunction).Function.PbCode(), Equals, tipb.ScalarFuncSig_StrToDateDatetime)
	exprs = append(exprs, function)

```

**Comment by JaySon-Huang** [comment]
Rely on https://github.com/pingcap/tidb/pull/23661, do not merge this PR.
Or rewrite this cherry-pick PR.

**Review by AilinKid — APPROVED** [review]

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**